kiwi_keg/generator.py

```diff
@@ -142,3 +155,19 @@ def _read_template(self, template_name):
                     location=self.description_schemas
                 )
             )
+
+    @staticmethod
+    def _preserve_XML_comments(text, xml_file):
+        comments = []
+        if text:
+            for comment in text.split(os.linesep):
+                if comment.startswith(('<!--', '-->', '<?')):
+                    comments.append(comment)
+        if comments:
+            with open(xml_file, 'r') as xml:
+                xml_data = xml.read()
+
+            with open(xml_file, 'w') as xml:
+                xml.write(os.linesep.join(comments))
+                xml.write(os.linesep)
+                xml.write(xml_data)
```

Comment:
Hmm, this is somewhat unrelated to this work but still something we might want to consider. At present our files have section comments such as <!-- initialization code --> looks like we loose that data. Given that what's in the build service will still be read by humans it would really be great to figure out a way to maintain those guiding comments as well.

Reply:
I'm looking into it

Reply:
Put back inline comments has many many problems. I tried almost all day to come up with a good solution and failed.

* You need an anchor in the document to put back an inline comment at the right place. line numbers does not work
   because they are different before and after the transformation

* The current implementation is super silly and just uses the previous line as anchor information but this might be wrong as the same information can exist at other places in the document. So worst case the comment is placed at the wrong place

* The XSLT transformation updates the schema. This means attributes and sections changes. If the anchor was an information that got changed by the transformation you will not find this position at all. So the comment gets lost

Overall this is the problem of diff/patch and there are very sophisticated algorithms to find the right place in the document but all that would be way too much overhead. Using patch is also not an option because it will fail if the XSLT transfornation has shifted the document to another schema.

I'm sorry my vote is to care for the header comments which can be placed safely and give up on inline comments

And another option would be to add a keg option that skips the schema update and allow users to do the move when it is required but not force this operation

Thoughts ?

Reply:
I will handle this via extra options

Reply:
I think the root of the issue is XSLT transforming the document, so fix should be there but that does not seem to be possible
I agree, better not to offer something that we know is going to be potentially wrong